app/workers/follows/update_points_worker.rb

```diff
@@ -0,0 +1,60 @@
+module Follows
+  class UpdatePointsWorker
+    include Sidekiq::Worker
+    sidekiq_options queue: :low_priority, retry: 10
+
+    def perform(article_id, user_id)
+      article = Article.find_by(id: article_id)
+      user = User.find_by(id: user_id)
+      return unless article && user
+
+      adjust_other_tag_follows_of_user(user.id)
+      followed_tag_names = user.cached_followed_tag_names
+      article.decorate.cached_tag_list_array.each do |tag_name|
+        if followed_tag_names.include?(tag_name)
+          recalculate_tag_follow_points(tag_name, user)
+        end
+      end
+    end
+
+    def recalculate_tag_follow_points(tag_name, user)
+      tag = Tag.find_by(name: tag_name)
+      follow = Follow.follower_tag(user.id).where(followable_id: tag.id).last
+
+      follow.implicit_points = calculate_implicit_points(tag, user)
+      follow.points = follow.implicit_points + follow.explicit_points
+      follow.save
+    end
+
+    def calculate_implicit_points(tag, user)
+      last_100_reactable_ids = user.reactions.where("points > 0").where(reactable_type: "Article")
+        .pluck(:reactable_id).last(100)
+      last_100_long_page_view_article_ids = user.page_views.where("time_tracked_in_seconds > 45")
+        .pluck(:article_id).last(100)
+      articles = Article.where(id: last_100_reactable_ids + last_100_long_page_view_article_ids)
+      tags = articles.pluck(:cached_tag_list).map { |list| list.split(", ") }.flatten
+      occurrences = tags.count(tag.name)
+      bonus = inverse_popularity_bonus(tag)
+      Math.log(occurrences + bonus + 1)
+    end
```

Comment:
My brain is kind of breaking trying to follow what this calculation does, exactly. 🤯  The log part makes sense, but why do we take the log of the occurrences and the bonus (I assume the `+ 1` is to ensure the number we take the log of is larger than 1, but I could be wrong)?Can we split it up or add some inline documentation? Since this is the _most_ important part of this worker, I think it's worth it to try to make its function as clear as possible

Reply:
@vaidehijoshi It's for the case a tag has 0 occurrences I guess, since `Math.log(0) #=> -Infinity`.

Reply:
Yeah, this is correct. I added a comment.